apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerDbType.cs

```diff
@@ -12,66 +12,165 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+using System.Collections.Generic;
+using Google.Cloud.Spanner.V1;
+using Google.Protobuf.Collections;
+
 namespace Google.Cloud.Spanner.Data
 {
     /// <summary>
     /// These types come from Google.Cloud.Spanner.V1.TypeCode
     /// </summary>
-    public enum SpannerDbType
+    public class SpannerDbType
     {
+        private readonly IDictionary<string, SpannerDbType> _structMembers;
+        private readonly TypeCode _typeCode;
+        private readonly SpannerDbType _elementType;
+
+        private SpannerDbType(TypeCode typeCode)
+        {
+            _typeCode = typeCode;
+        }
+
+        private SpannerDbType(TypeCode typeCode, SpannerDbType arrayElementType)
+            :this(typeCode)
+        {
+            _elementType = arrayElementType;
+        }
+
+        private SpannerDbType(TypeCode typeCode, IDictionary<string, SpannerDbType> structMembers)
+            : this(typeCode)
+        {
+            _structMembers = structMembers;
+        }
+
+        private SpannerDbType(TypeCode typeCode, RepeatedField<StructType.Types.Field> structTypeFields)
+            : this(typeCode)
+        {
+            _structMembers = new Dictionary<string, SpannerDbType>();
+            foreach (StructType.Types.Field field in structTypeFields)
+            {
+                _structMembers[field.Name] = FromV1Type(field.Type);
+            }
+        }
+
+        internal static SpannerDbType FromV1Type(V1.Type type)
```

Comment:
Possibly FromProtobufType so that the name doesn't need to change when we move to V2 in the distant future?